### PR TITLE
Wire live updates to package metadata view service [DPP-1154]

### PR DESCRIPTION
**Pull request by skisel-da**
CHANGELOG_BEGIN
CHANGELOG_END

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/index/InMemoryStateUpdater.scala [hunk 1]
```diff
@@ -44,12 +46,21 @@ final class InMemoryStateUpdater(
...
+        .collect { case (_, event: Update.PublicPackageUpload) => event }
+        .foreach(updatePackageMetadata)
+  }
+
   // TODO LLP: Considering directly returning this flow instead of the wrapper
-  def flow: UpdaterFlow =
+  val flow: UpdaterFlow =
     Flow[(Vector[(Offset, Update)], Long)]
       .filter(_._1.nonEmpty)
+      .alsoTo(packageMetadataFlow)
```
Comment (nmarton-da) [inline ledger/participant-integration-api/src/main/scala/platform/index/InMemoryStateUpdater.scala:63]: AFAIK alsoTo will break consistency for us.
Our requirement is, that as ledger end is updated, we guarantee that the pacakge metadata is updated upto ledger end.
This alsoTo will start a computation which will indeed backpressure, but the same batch will flow through the main pipeline without any delay, which at the end bumps the ledger end, and announces the new ledger end for the streams (bumps the Dispatcher).
So these guys are racing.
But ledger end pumping must come after updating the packageMetadataView

File: ledger/participant-integration-api/src/main/scala/platform/index/InMemoryStateUpdater.scala [hunk 2]
```diff
@@ -44,12 +46,21 @@ final class InMemoryStateUpdater(
...
     ) => TransactionLogUpdate.TransactionRejected,
     updateLedgerEnd: (Offset, Long) => Unit,
+    updatePackageMetadata: Update.PublicPackageUpload => Unit,
 ) {
 
+  val packageMetadataFlow: Sink[(Vector[(Offset, Update)], Long), Future[Done]] = Sink.foreach {
+    case (batch, _) =>
+      batch
+        .collect { case (_, event: Update.PublicPackageUpload) => event }
+        .foreach(updatePackageMetadata)
```
Comment (nmarton-da) [inline ledger/participant-integration-api/src/main/scala/platform/index/InMemoryStateUpdater.scala:56]: We are running here expensive computation on the akka default scheduler. That is not good, we want to run the computation on an appropriate and dedicated execution context

File: ledger/participant-integration-api/src/main/scala/platform/index/InMemoryStateUpdater.scala [hunk 3]
```diff
@@ -100,14 +93,42 @@ private[platform] object InMemoryStateUpdater {
...
-    convertTransactionAccepted = convertTransactionAccepted,
-    convertTransactionRejected = convertTransactionRejected,
-    updateLedgerEnd = updateLedgerEnd(inMemoryState),
+    prepare = prepare,
+    update = update(inMemoryState, loggingContext),
   )
 
-  private def updateCaches(inMemoryState: InMemoryState)(
-      updates: Vector[TransactionLogUpdate]
+  def combinePackageMetadata(batch: Vector[(Offset, Update)]): PackageMetadata =
```
Comment (nmarton-da) [inline ledger/participant-integration-api/src/main/scala/platform/index/InMemoryStateUpdater.scala:100]: naming nit: naming suggest in this form this is about some aggregation, but in fact we decode and do the heavy lifting in the middle, can we reflect this in naming somehow?

File: ledger/participant-integration-api/src/main/scala/platform/index/InMemoryStateUpdater.scala [hunk 4]
```diff
@@ -100,14 +93,42 @@ private[platform] object InMemoryStateUpdater {
...
+      packageMetadata = combinePackageMetadata(batch),
+    )
+
+  def update(
+      inMemoryState: InMemoryState,
+      loggingContext: LoggingContext,
+  )(result: PrepareResult): Unit = {
+    inMemoryState.packageMetadataView.update(result.packageMetadata)
+    updateCaches(inMemoryState, result.logUpdate)
+    updateLedgerEnd(inMemoryState, result.lastOffset, result.lastEventSequentialId)(loggingContext)
```
Comment (nmarton-da) [inline ledger/participant-integration-api/src/main/scala/platform/index/InMemoryStateUpdater.scala:126]: please add a comment here that this needs to come last in this method and the reason for it

**Review by nmarton-da — APPROVED** [review]
I like this refactoring :raised_hands: 
@tudor-da can you also please have a look?

**Comment by nmarton-da** [comment]
@skisel-da the spec test is failing on CI

File: ledger/participant-integration-api/src/main/scala/platform/index/InMemoryStateUpdater.scala (hunk 4 above)
Comment (skisel-da) [reply ledger/participant-integration-api/src/main/scala/platform/index/InMemoryStateUpdater.scala:126]: there is such a comment already inside `updateLedgerEnd` i will add another one on top of it

File: ledger/participant-integration-api/src/main/scala/platform/index/InMemoryStateUpdater.scala [hunk 5]
```diff
@@ -34,43 +35,35 @@ final class InMemoryStateUpdater(
...
-          metrics.daml.index.ledgerEndSequentialId.updateValue(lastEventSequentialId)
+          update(result)
+          metrics.daml.index.ledgerEndSequentialId.updateValue(result.lastEventSequentialId)
         }(updateCachesExecutionContext)
       }
 }
 
 private[platform] object InMemoryStateUpdater {
+  case class PrepareResult(
+      logUpdate: Vector[TransactionLogUpdate],
```
Comment (tudor-da) [inline ledger/participant-integration-api/src/main/scala/platform/index/InMemoryStateUpdater.scala:62]:
```suggestion
      updates: Vector[TransactionLogUpdate],
```

File: ledger/participant-integration-api/src/test/suite/scala/platform/index/InMemoryStateUpdaterSpec.scala [hunk 6]
```diff
@@ -51,15 +49,17 @@ class InMemoryStateUpdaterSpec
...
-    runFlow(Seq(Vector(update1, metadataChangedUpdate) -> 1L, Vector(update3, update4) -> 3L))
-
-    cacheUpdates should contain theSameElementsInOrderAs Seq(
-      Vector(txLogUpdate1),
-      Vector(txLogUpdate3, txRejected),
+    runFlow(
+      Seq(
+        Vector(update1, metadataChangedUpdate) -> 1L,
+        Vector(update3, update4) -> 3L,
+        Vector(update5) -> 4L,
```
Comment (tudor-da) [inline ledger/participant-integration-api/src/test/suite/scala/platform/index/InMemoryStateUpdaterSpec.scala:56]: It's not visible in the assertion what happens with these vectors or their contents, but only the evt seq id. Can you make sure we don't lose entropy here?

File: ledger/participant-integration-api/src/test/suite/scala/platform/index/InMemoryStateUpdaterSpec.scala [hunk 7]
```diff
@@ -72,53 +72,36 @@ class InMemoryStateUpdaterSpec
         // Results in empty batch after processing
         // Should still have effect on ledger end updates
         Vector(anotherMetadataChangedUpdate) -> 3L,
+        Vector(update5) -> 4L,
```
Comment (tudor-da) [inline ledger/participant-integration-api/src/test/suite/scala/platform/index/InMemoryStateUpdaterSpec.scala:75]: same comment as above

File: ledger/participant-integration-api/src/main/scala/platform/index/InMemoryStateUpdater.scala [hunk 8]
```diff
@@ -100,14 +93,43 @@ private[platform] object InMemoryStateUpdater {
...
   )
 
-  private def updateCaches(inMemoryState: InMemoryState)(
-      updates: Vector[TransactionLogUpdate]
+  def extractMetadataFromUploadedPackage(batch: Vector[(Offset, Update)]): PackageMetadata =
+    batch.foldLeft(PackageMetadata()) {
+      case (metadata, (_, uploadEvent: Update.PublicPackageUpload)) =>
+        uploadEvent.archives.foldLeft(metadata) { case (acc, archive) =>
+          acc.append(PackageMetadata.from(archive))
+        }
```
Comment (tudor-da) [inline ledger/participant-integration-api/src/main/scala/platform/index/InMemoryStateUpdater.scala:105]:
```suggestion
        uploadEvent.archives.view
          .map(PackageMetadata.from)
          .foldLeft(metadata)(_ append _)
```

File: ledger/participant-integration-api/src/main/scala/platform/index/InMemoryStateUpdater.scala [hunk 9]
```diff
@@ -100,14 +93,43 @@ private[platform] object InMemoryStateUpdater {
...
+      case (metadata, _) => metadata
+    }
```
Comment (tudor-da) [inline ledger/participant-integration-api/src/main/scala/platform/index/InMemoryStateUpdater.scala:107]:
```suggestion
  def extractMetadataFromUploadedPackage(batch: Vector[(Offset, Update)]): PackageMetadata =
    batch.view
      .collect { case (_, packageUpload: Update.PublicPackageUpload) => packageUpload }
      .flatMap(_.archives.view)
      .map(PackageMetadata.from)
      .foldLeft(PackageMetadata())(_ append _)
```

File: ledger/participant-integration-api/src/main/scala/platform/index/InMemoryStateUpdater.scala [hunk 10]
```diff
@@ -100,14 +93,43 @@ private[platform] object InMemoryStateUpdater {
...
+
+  def prepare(batch: Vector[(Offset, Update)], lastEventSequentialId: Long): PrepareResult =
+    PrepareResult(
+      logUpdate = batch.collect {
+        case (offset, u: Update.TransactionAccepted) => convertTransactionAccepted(offset, u)
+        case (offset, u: Update.CommandRejected) => convertTransactionRejected(offset, u)
+      },
+      lastOffset = batch.last._1,
+      lastEventSequentialId = lastEventSequentialId,
+      packageMetadata = extractMetadataFromUploadedPackage(batch),
```
Comment (tudor-da) [inline ledger/participant-integration-api/src/main/scala/platform/index/InMemoryStateUpdater.scala:117]:
```suggestion
      packageMetadata = extractMetadataFromUploadedPackages(batch),
```

File: ledger/participant-integration-api/src/main/scala/platform/index/InMemoryStateUpdater.scala (hunk 9 above)
Comment (skisel-da) [reply ledger/participant-integration-api/src/main/scala/platform/index/InMemoryStateUpdater.scala:107]: this is way much nicer than my peace, thanks!

File: ledger/participant-integration-api/src/test/suite/scala/platform/index/InMemoryStateUpdaterSpec.scala (hunk 6 above)
Comment (skisel-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/index/InMemoryStateUpdaterSpec.scala:56]: yeah I switched this PR to draft before I wanted to make the test green :) Need to find a way here

Comment (skisel-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/index/InMemoryStateUpdaterSpec.scala:60]: I've made an attempt, please have a look

**Review by tudor-da — APPROVED** [review]

**Pull request merged**